### PR TITLE
Update include documentation #2

### PR DESCRIPTION
**Pull request by Nextra**
Second grind at our outdated/non-existent/inconsistent include documentation (see #108). Will start in alphabetical order and see when I get tired of doing it :)

Early work in progress PR, feedback welcomed. ~~Documentation ideas for `constraint_offset` in amxmisc.inc very welcomed.~~

Thanks to @xPaw, @Freeman-AM and of course @Arkshine for helping out :sun_with_face:

### CODE REVIEW

**Comment by ClaudiuHKS** [comment]
Not bad. What is WIP?

**Comment by Nextra** [comment]
> Early **work in progress** PR

**Comment by Nextra** [comment]
core.inc added with the risk of containing some grave mistakes :disappointed:

**Comment by Freeman-AM** [comment]
Lets talk a bit about native and forward type of writing style.

Basically there is 2 styles used:

1: native ArraySortEx(Array:array, const comparefunc[], data[]="", data_size=0)
Used randomly but mostly in amxmodx.inc

2: native ArraySortEx(Array:array, const comparefunc[], data[] = "", data_size = 0)
Used randomly.

Example of bad formating : ArraySortEx( Array:array , const comparefunc[] ,data[]="", data_size = 0)

I think we should keep the same style everywhere for every inc.
So lets vote witch style do you prefer.

For me i think that 2 is more readable, it's also Arkshine point of view.

**Comment by Nextra** [comment]
It should be 2, and I already made some of those changes. I'll update it anywhere I notice, but this could also be a thing for a followup PR.

**Comment by Nextra** [comment]
Thanks to @Freeman-AM for helping me go through amxmodx.inc again, finding a lot of typos and mistakes I made the last time around.

**Comment by ClaudiuHKS** [comment]
Oh come on, Freeman... Voting for INC style?
I saw in my life a thousand of editors that complete and/ or fix your code automatically...
A few of them add tabs, a few of them add 2-3 spaces, other ones add 5-6 spaces.

Now, while the code works like a charm, who cares if ="" or = ""?
If you guys really care of the visual style, why not edit all AMX Mod X files to look the same?

I saw that 20% of code is aligned with spaces and 80% with tabs. Anyways, Microsoft Visual Studio 2013 and 2015 Preview align everything the same way when copy-pasted. Even 10,000 lines file.

Too much headache. But, I'd like to see the same style in every file of AMX Mod X. Can't disguise that.

**Comment by xPaw** [comment]
@ClaudiuHKS It's a good idea to unify the style while there are changes being made to inc files.

**Comment by Freeman-AM** [comment]
* About includes:
It's not like people writing .inc everydays.
Style formatting can be done by another person after the release of the .inc.

* About amxmodx plugins:
Almost all amxmodx plugins need a rewrite for minor improvements and sometimes for major improvements (plmenu).
Fixing the style will be done when we will attack this part.
Of course we know how to code, we know how to write and so on...
But i think that readability is an important scope.
We are not alone, everybody read .inc and .sma.
If it's badly written for use, it's even worse for beginners or intermediate users.

**Comment by Nextra** [comment]
csstats.inc and csx.inc with a high chance of errors, ewwwwww.

**Comment by Nextra** [comment]
New updates. With all the dynamic datastructures and our most important mod done I think this should wrap it up. I'll tackle the engine/fakemeta/message stuff (which is next on my todo) in a followup PR. Please get all your feedback in soon-ish so we can get this out there and into the docgen, thanks!

Note for the Datapack include: Discussions on IRC lead to the decision to nuke the IsPackReadable native, more on that in the PR that actually removes it internally.

File: plugins/include/csstats.inc [hunk 1]
```diff
@@ -8,62 +8,267 @@
...
+native get_user_wrstats(index, wpnindex, stats[8], bodyhits[8]);
 
-/* Gets round stats from given weapon index.*/
-native get_user_wrstats(index,wpnindex,stats[8],bodyhits[8]);
+/**
+ * Retrieves the client's weapon statistics from the permanent storage on the
+ * server.
+ *
+ * @note The permanent storage is updated on every respawn or client disconnect.
+ * @note Player rank is determined by the difference of kills to deaths.
```
Comment (Arkshine) [inline plugins/include/csstats.inc]: This is kind of imprecise. Position calculation is provided by a custom function you can modify (csstats.sma). This is probably worth to explain that.

File: plugins/include/csstats.inc [hunk 2]
```diff
@@ -8,62 +8,267 @@
...
-/* Gets overall stats which stored in stats.dat file in amx folder
-* and updated on every mapchange or user disconnect.
-* Function returns next index of stats entry or 0 if no more exists. */
-native get_stats(index,stats[8],bodyhits[8],name[],len,authid[] = "",authidlen = 0);
+/**
+ * Retrieves statistics from the permanent storage on the server via iterative,
+ * incremental access.
+ *
+ * @note The permanent storage is updated on every respawn or client disconnect.
+ * @note Player rank is determined by the difference of kills to deaths.
```
Comment (Arkshine) [inline plugins/include/csstats.inc]: Same here.

File: plugins/include/cstrike.inc [hunk 3]
```diff
@@ -25,329 +25,1060 @@
...
  */
 native cs_get_user_bpammo(index, weapon);
 
-/* Restock/remove ammo in a user's backpack.
+/**
+ * Sets amount of ammo in the client's backpack for a specific weapon.
+ *
+ * @note For a list of possible weapon ids see the CSW_* constants in
+ *       amxconst.inc
+ * @note Some weapons share ammo types and therefore ammo backpack pools. List
```
Comment (Arkshine) [inline plugins/include/cstrike.inc]: Looks like you did not align like in cs_get_user_bpammo

File: plugins/include/cstrike.inc [hunk 4]
```diff
@@ -25,329 +25,1060 @@
...
+ * @error           If the provided entity index is not a bomb, an error will be
+ *                  thrown.
+ */
 native cs_set_c4_explode_time(index, Float:value);
 
-/* Gets or sets whether the C4 is being defused.
+/**
+ * Returns if the bomb is being defused.
+ *
+ * @param index     C4 entity
```
Comment (Arkshine) [inline plugins/include/cstrike.inc]: Should be "c4index"

File: plugins/include/cstrike.inc [hunk 5]
```diff
@@ -25,329 +25,1060 @@
...
+ * @return          1 if the bomb is being defused, 0 otherwise
+ * @error           If the provided entity index is not a bomb, an error will be
+ *                  thrown.
  */
 native bool:cs_get_c4_defusing(c4index);
 
+/**
+ * Sets if the bomb is being defused.
+ *
+ * @param index     C4 entity
```
Comment (Arkshine) [inline plugins/include/cstrike.inc]: Same here

File: plugins/include/csx.inc [hunk 6]
```diff
@@ -27,65 +27,237 @@
...
+ *       amxconst.inc
+ * @note For a list of possible body hitplaces see the HIT_* constants in
+ *       amxconst.inc
+ *
+ * @param attacker  Attacker client index
+ * @param victim    Victim client index
+ * @param damage    Damage dealt to victim
+ * @param wpnindex  Weapon id
+ * @param hitplace  Body hitplace
+ * @param ta        If nonzero the attack was a team attack
```
Comment (Arkshine) [inline plugins/include/csx.inc]: should be "TA",

File: plugins/include/csx.inc [hunk 7]
```diff
@@ -27,65 +27,237 @@
...
+ * @param melee     If nonzero the weapon will be considered a melee weapon
+ * @param logname   Weapon short name
+ *
+ * @return          Cusom weapon id (>0) on success, 0 if no more custom weapons
+ *                  can be added
+ */
+native custom_weapon_add(const wpnname[], melee = 0, const logname[] = "");
+
+/**
+ * Trigger a damage event on a custom weapon, adding it to the internal stats.
```
Comment (Arkshine) [inline plugins/include/csx.inc]: Triggers?

File: plugins/include/csx.inc [hunk 8]
```diff
@@ -27,65 +27,237 @@
...
+ * Add a shot event on a custom weapon to the internal stats.
+ *
+ * @param weapon    Custom weapon id
+ * @param index     Client index
+ *
+ * @noreturn
+ * @error           If the weapon id is not a custom weapon or an invalid client
+ *                  index is provided, an error will be thrown.
+ */
+native custom_weapon_shot(weapon, index); // weapon id , player id
```
Comment (Arkshine) [inline plugins/include/csx.inc]: Comment sounds like unecessary

File: plugins/include/csx.inc [hunk 9]
```diff
@@ -27,65 +27,237 @@
...
+ * @noreturn
+ * @error           If the weapon id is not a custom weapon, an invalid client
+ *                  index, damage value or hitplace is provided, an error will
+ *                  be thrown.
+ */
+native custom_weapon_dmg(weapon, att, vic, damage, hitplace = 0);
 
-/* function will return 1 if true */
+/**
+ * Add a shot event on a custom weapon to the internal stats.
```
Comment (Arkshine) [inline plugins/include/csx.inc]: Adds?

File: plugins/include/cvars.inc [hunk 10]
```diff
@@ -93,56 +113,59 @@ native get_cvar_pointer(const cvar[]);
...
- * @error               Invalid hook handle
+ * @param handle    Forward to enable
+ * @error           If an invalid hook handle is provided, an error will be
+ *                  thrown.
  */
 native enable_cvar_hook(cvarhook:handle);
 
 /**
- * Returns flags of a cvar. The cvar is accessed by name.
+ * Returns a flags of a cvar. The cvar is accessed by name.
```
Comment (Arkshine) [inline plugins/include/cvars.inc]: "a flags", probably not the best phrased?

File: plugins/include/cvars.inc [hunk 11]
```diff
@@ -93,56 +113,59 @@ native get_cvar_pointer(const cvar[]);
...
  *
  * @note For a list of possible flags see the FCVAR_* constants in amxconst.inc
- * @note Accessing a Cvar by name requires this function to walk through the
- *       engine's cvar list every time, which can result in a considerable waste
- *       of processing time, especially if many cvars have been registered. For
- *       a vastly superior alternative look at the get_pcvar_flags function.
+ * @note Accessing a Cvar by name is slower than direct pointer access, which is
```
Comment (Arkshine) [inline plugins/include/cvars.inc]: Well, such cvars are now relying on the cache, so technically it's true the first time, but once it's cached, not really. Well, it's fine to leave as it is though, but could be okay to add some note about it maybe?

**Comment by Arkshine** [comment]
Well probably missed things but looks good.

File: plugins/include/csstats.inc (hunk 1 above)
Comment (Nextra) [inline plugins/include/csstats.inc]: I never knew that existed. I'm going to think of something.

File: plugins/include/csx.inc (hunk 8 above)
Comment (Nextra) [inline plugins/include/csx.inc]: Yup, missed that.

File: plugins/include/cvars.inc (hunk 10 above)
Comment (Nextra) [inline plugins/include/cvars.inc]: Good catch

**Comment by Jhob94** [comment]
Maybe add this into engine_stocks.inc? http://pastebin.com/24cTbrMM

**Pull request merged**